File: android/app/src/main/java/com/transcribecare/app/service/SpeechRecognitionService.kt
package com.transcribecare.app.service

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.speech.RecognitionListener
import android.speech.RecognizerIntent
import android.speech.SpeechRecognizer

/**
 * Wrapper around Android's SpeechRecognizer that provides continuous speech-to-text
 * recognition with interim and final results, auto-restart on unexpected session end,
 * and graceful error handling.
 *
 * @param context Application context used to create the SpeechRecognizer instance.
 * @param onPartialResult Callback invoked with interim (non-final) recognition text.
 * @param onFinalResult Callback invoked with finalized recognition text.
 * @param onError Callback invoked with a user-facing error message.
 */
class SpeechRecognitionService(
    private val context: Context,
    private val onPartialResult: (interimText: String) -> Unit,
    private val onFinalResult: (text: String) -> Unit,
    private val onError: (message: String) -> Unit
) {

    private var speechRecognizer: SpeechRecognizer? = null
    private var isRecordingIntent: Boolean = false
    private val handler = Handler(Looper.getMainLooper())

    companion object {
        private const val RETRY_DELAY_MS = 500L
    }

    /**
     * Starts continuous speech recognition. Creates a new SpeechRecognizer instance
     * and begins listening with partial results enabled.
     */
    fun startListening() {
        isRecordingIntent = true
        initializeRecognizer()
        speechRecognizer?.startListening(createRecognizerIntent())
    }

    /**
     * Stops speech recognition. Prevents auto-restart by clearing the recording
     * intent flag. The recognizer is not destroyed immediately — it will be
     * destroyed after delivering its final results via onResults, or on the next
     * restartIfActive() call which will see isRecordingIntent is false.
     */
    fun stopListening() {
        isRecordingIntent = false
        speechRecognizer?.stopListening()
    }

    /**
     * Releases the SpeechRecognizer and clears all pending handler callbacks.
     * Should be called when the service is no longer needed (e.g., ViewModel onCleared).
     */
    fun destroy() {
        isRecordingIntent = false
        handler.removeCallbacksAndMessages(null)
        destroyRecognizer()
    }

    private fun initializeRecognizer() {
        destroyRecognizer()
        speechRecognizer = SpeechRecognizer.createSpeechRecognizer(context).apply {
            setRecognitionListener(createRecognitionListener())
        }
    }

    private fun destroyRecognizer() {
        speechRecognizer?.destroy()
        speechRecognizer = null
    }

    private fun createRecognizerIntent(): Intent {
        return Intent(RecognizerIntent.ACTION_RECOGNIZE_SPEECH).apply {
            putExtra(
                RecognizerIntent.EXTRA_LANGUAGE_MODEL,
                RecognizerIntent.LANGUAGE_MODEL_FREE_FORM
            )
            putExtra(RecognizerIntent.EXTRA_PARTIAL_RESULTS, true)
        }
    }

    private fun createRecognitionListener(): RecognitionListener {
        return object : RecognitionListener {

            override fun onReadyForSpeech(params: Bundle?) {
                // Recognition is ready — no action needed
            }

            override fun onBeginningOfSpeech() {
                // User started speaking — no action needed
            }

            override fun onRmsChanged(rmsdB: Float) {
                // Audio level changed — could be used for visual feedback
            }

            override fun onBufferReceived(buffer: ByteArray?) {
                // Raw audio buffer — not used
            }

            override fun onEndOfSpeech() {
                // User stopped speaking — results will follow via onResults
            }

            override fun onPartialResults(partialResults: Bundle?) {
                val matches = partialResults
                    ?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
                val interimText = matches?.firstOrNull().orEmpty()
                if (interimText.isNotEmpty()) {
                    onPartialResult(interimText)
                }
            }

            override fun onResults(results: Bundle?) {
                val matches = results
                    ?.getStringArrayList(SpeechRecognizer.RESULTS_RECOGNITION)
                val finalText = matches?.firstOrNull().orEmpty()
                if (finalText.isNotEmpty()) {
                    onFinalResult(finalText)
                }
                // Auto-restart to maintain continuous recognition
                restartIfActive()
            }

            override fun onError(error: Int) {
                handleError(error)
            }

            override fun onEvent(eventType: Int, params: Bundle?) {
                // Reserved for future use by the framework
            }
        }
    }

    /**
     * Handles SpeechRecognizer error codes with appropriate recovery strategies:
     * - ERROR_NO_MATCH: Continue listening silently (no user-facing error)
     * - ERROR_NETWORK: Display warning, attempt to continue
     * - ERROR_RECOGNIZER_BUSY: Retry after a short delay
     * - Other errors: Stop recognition and report to user
     */
    private fun handleError(errorCode: Int) {
        when (errorCode) {
            SpeechRecognizer.ERROR_NO_MATCH -> {
                // No speech detected — continue listening without notifying user
                restartIfActive()
            }

            SpeechRecognizer.ERROR_SPEECH_TIMEOUT -> {
                // Speech input timed out — restart if still recording
                restartIfActive()
            }

            SpeechRecognizer.ERROR_NETWORK,
            SpeechRecognizer.ERROR_NETWORK_TIMEOUT -> {
                onError("Network unavailable. Attempting to continue...")
                restartIfActive()
            }

            SpeechRecognizer.ERROR_RECOGNIZER_BUSY -> {
                // Recognizer is busy — retry after a short delay
                handler.postDelayed({
                    restartIfActive()
                }, RETRY_DELAY_MS)
            }

            SpeechRecognizer.ERROR_CLIENT -> {
                // Client-side error — attempt restart if recording
                restartIfActive()
            }

            SpeechRecognizer.ERROR_INSUFFICIENT_PERMISSIONS -> {
                isRecordingIntent = false
                onError("Microphone permission is required for speech recognition.")
                destroyRecognizer()
            }

            SpeechRecognizer.ERROR_SERVER -> {
                onError("Speech recognition server error. Please try again.")
                isRecordingIntent = false
                destroyRecognizer()
            }

            else -> {
                onError("Speech recognition error. Please try again.")
                isRecordingIntent = false
                destroyRecognizer()
            }
        }
    }

    /**
     * Restarts recognition if the user's recording intent is still active.
     * This enables continuous transcription across recognition sessions.
     * If recording is no longer intended, destroys the recognizer to free resources.
     */
    private fun restartIfActive() {
        if (isRecordingIntent) {
            initializeRecognizer()
            speechRecognizer?.startListening(createRecognizerIntent())
        } else {
            destroyRecognizer()
        }
    }
}

File: ios/TranscribeCare/Service/SpeechRecognitionService.swift
import Foundation
import Speech
import AVFoundation

/// Wrapper around SFSpeechRecognizer that provides continuous on-device speech recognition
/// with interim and final results, auto-restart on unexpected session end, and graceful
/// error handling with on-device fallback.
///
/// Uses AVAudioEngine to tap the audio input node and feed buffers to the recognition request.
///
/// - Requirements: 4.1, 4.4, 4.5, 4.7, 4.8
@Observable
class SpeechRecognitionService {

    // MARK: - Callbacks

    /// Called with interim (non-final) recognition text while the user is speaking.
    var onPartialResult: ((String) -> Void)?

    /// Called with finalized recognition text when the recognizer commits a result.
    var onFinalResult: ((String) -> Void)?

    /// Called with a user-facing error message when recognition encounters a problem.
    var onError: ((String) -> Void)?

    // MARK: - Public State

    /// Whether the service is actively listening for speech.
    private(set) var isListening: Bool = false

    // MARK: - Private Properties

    private let speechRecognizer: SFSpeechRecognizer?
    private let audioEngine = AVAudioEngine()
    private var recognitionRequest: SFSpeechAudioBufferRecognitionRequest?
    private var recognitionTask: SFSpeechRecognitionTask?

    /// Tracks whether the user intends to be recording. Used to determine
    /// if auto-restart should occur when a session ends unexpectedly.
    private var isRecordingIntent: Bool = false

    // MARK: - Initialization

    /// Creates a new SpeechRecognitionService with the default locale recognizer.
    init() {
        self.speechRecognizer = SFSpeechRecognizer(locale: Locale.current)
    }

    // MARK: - Public Methods

    /// Starts continuous speech recognition.
    ///
    /// Configures the audio session, sets up the audio engine input tap,
    /// creates a recognition request, and begins processing speech.
    func startListening() {
        // Check authorization status
        guard checkAuthorization() else { return }

        // Verify recognizer is available
        guard let speechRecognizer = speechRecognizer, speechRecognizer.isAvailable else {
            onError?("Speech recognition is not available on this device.")
            return
        }

        isRecordingIntent = true
        isListening = true

        do {
            try startRecognitionSession()
        } catch {
            isListening = false
            isRecordingIntent = false
            onError?("Failed to start speech recognition: \(error.localizedDescription)")
        }
    }

    /// Stops speech recognition and releases resources.
    /// Prevents auto-restart by clearing the recording intent flag.
    func stopListening() {
        isRecordingIntent = false
        isListening = false
        stopRecognitionSession()
    }

    /// Releases all resources. Call when the service is no longer needed.
    func destroy() {
        isRecordingIntent = false
        isListening = false
        stopRecognitionSession()
    }

    // MARK: - Private Methods

    /// Checks speech recognition authorization and reports errors for denied states.
    /// - Returns: `true` if authorized, `false` otherwise.
    private func checkAuthorization() -> Bool {
        let status = SFSpeechRecognizer.authorizationStatus()
        switch status {
        case .authorized:
            return true
        case .denied:
            onError?("Speech recognition permission denied. Please enable it in Settings > Privacy > Speech Recognition.")
            return false
        case .restricted:
            onError?("Speech recognition is restricted on this device.")
            return false
        case .notDetermined:
            // Request authorization and report result asynchronously
            requestAuthorization()
            return false
        @unknown default:
            onError?("Speech recognition authorization status unknown.")
            return false
        }
    }

    /// Requests speech recognition authorization from the user.
    private func requestAuthorization() {
        SFSpeechRecognizer.requestAuthorization { [weak self] status in
            DispatchQueue.main.async {
                switch status {
                case .authorized:
                    self?.startListening()
                case .denied:
                    self?.onError?("Speech recognition permission denied. Please enable it in Settings > Privacy > Speech Recognition.")
                case .restricted:
                    self?.onError?("Speech recognition is restricted on this device.")
                case .notDetermined:
                    self?.onError?("Speech recognition authorization not determined.")
                @unknown default:
                    self?.onError?("Speech recognition authorization status unknown.")
                }
            }
        }
    }

    /// Configures and starts a recognition session with audio engine input.
    private func startRecognitionSession() throws {
        // Cancel any existing task
        stopRecognitionSession()

        // Configure audio session for recording
        let audioSession = AVAudioSession.sharedInstance()
        try audioSession.setCategory(.record, mode: .measurement, options: .duckOthers)
        try audioSession.setActive(true, options: .notifyOthersOnDeactivation)

        // Create recognition request
        recognitionRequest = SFSpeechAudioBufferRecognitionRequest()
        guard let recognitionRequest = recognitionRequest else {
            throw SpeechRecognitionError.requestCreationFailed
        }

        // Enable partial results for interim text display
        recognitionRequest.shouldReportPartialResults = true

        // Set on-device recognition as fallback for network issues (Requirement 4.7)
        if #available(iOS 13, *) {
            recognitionRequest.requiresOnDeviceRecognition = true
        }

        // Install audio tap on the input node
        let inputNode = audioEngine.inputNode
        let recordingFormat = inputNode.outputFormat(forBus: 0)
        inputNode.installTap(onBus: 0, bufferSize: 1024, format: recordingFormat) { [weak self] buffer, _ in
            self?.recognitionRequest?.append(buffer)
        }

        // Start audio engine
        audioEngine.prepare()
        try audioEngine.start()

        // Start recognition task
        guard let speechRecognizer = speechRecognizer else {
            throw SpeechRecognitionError.recognizerUnavailable
        }

        recognitionTask = speechRecognizer.recognitionTask(with: recognitionRequest) { [weak self] result, error in
            self?.handleRecognitionResult(result: result, error: error)
        }
    }

    /// Handles recognition results and errors from the recognition task.
    private func handleRecognitionResult(result: SFSpeechRecognitionResult?, error: Error?) {
        var isFinal = false

        if let result = result {
            let transcription = result.bestTranscription.formattedString
            isFinal = result.isFinal

            if isFinal {
                onFinalResult?(transcription)
            } else {
                onPartialResult?(transcription)
            }
        }

        if let error = error {
            handleRecognitionError(error)
            return
        }

        // If the result is final, the session has ended — auto-restart if intent is active
        if isFinal {
            restartIfActive()
        }
    }

    /// Handles recognition errors with appropriate recovery strategies.
    private func handleRecognitionError(_ error: Error) {
        let nsError = error as NSError

        // Check for specific error conditions
        if nsError.domain == "kAFAssistantErrorDomain" {
            switch nsError.code {
            case 1: // Recognition request was canceled
                restartIfActive()
                return
            case 4: // No speech detected / timeout
                restartIfActive()
                return
            default:
                break
            }
        }

        // Network-related errors: attempt on-device fallback
        if nsError.code == 2 || nsError.localizedDescription.lowercased().contains("network") {
            onError?("Network unavailable. Using on-device recognition.")
            restartIfActive()
            return
        }

        // For other errors, stop and report
        stopRecognitionSession()
        isListening = false

        if isRecordingIntent {
            // Unexpected end — attempt restart
            isRecordingIntent = true
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
                self?.restartIfActive()
            }
        } else {
            onError?("Speech recognition error: \(error.localizedDescription)")
        }
    }

    /// Stops the current recognition session and cleans up resources.
    private func stopRecognitionSession() {
        // Stop audio engine and remove tap
        if audioEngine.isRunning {
            audioEngine.stop()
            audioEngine.inputNode.removeTap(onBus: 0)
        }

        // End recognition request
        recognitionRequest?.endAudio()
        recognitionRequest = nil

        // Cancel recognition task
        recognitionTask?.cancel()
        recognitionTask = nil
    }

    /// Restarts recognition if the user's recording intent is still active.
    /// This enables continuous transcription across recognition sessions (Requirement 4.8).
    private func restartIfActive() {
        guard isRecordingIntent else { return }

        stopRecognitionSession()

        DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) { [weak self] in
            guard let self = self, self.isRecordingIntent else { return }
            do {
                try self.startRecognitionSession()
            } catch {
                self.isListening = false
                self.isRecordingIntent = false
                self.onError?("Failed to restart speech recognition: \(error.localizedDescription)")
            }
        }
    }
}

// MARK: - Error Types

/// Internal errors for the speech recognition service.
private enum SpeechRecognitionError: LocalizedError {
    case requestCreationFailed
    case recognizerUnavailable

    var errorDescription: String? {
        switch self {
        case .requestCreationFailed:
            return "Failed to create speech recognition request."
        case .recognizerUnavailable:
            return "Speech recognizer is not available."
        }
    }
}
